system-addon/lib/TippyTopFeed.jsm

```diff
@@ -0,0 +1,154 @@
+
+this.TippyTopFeed = class TippyTopFeed {
+  constructor() {
+    this.tippyTopLastUpdated = 0;
+    this.cache = new PersistentCache("tippytop", true);
+    this.prefs = new Prefs();
+    this.initialized = false;
+    this.initializing = false;
+    this.fetchQueue = [];
+  }
+
+  async init() {
+    this.initializing = true;
+    await this.loadCachedData();
+    await this.maybeRefresh();
+    this.initialized = true;
+    this.initializing = false;
+
+    while (this.fetchQueue.length > 0) {
+      this.fetchIcon(this.fetchQueue.pop());
+    }
+  }
+
+  async maybeRefresh() {
+    if (Date.now() - this.tippyTopLastUpdated >= TIPPYTOP_UPDATE_TIME) {
+      await this.refresh();
+    }
+  }
+
+  async refresh() {
+    let headers = new Headers();
+    if (this.etag) {
+      headers.set("If-None-Match", this.etag);
+    }
+    let {data, etag, status} = await this.loadFromURL(this.prefs.get("tippyTop.service.endpoint"), headers);
+    if (status === 304) {
+      // Not modified, we are done. No need to dispatch actions or update cache. We'll check again in 24+ hours.
+      this.tippyTopLastUpdated = Date.now();
+      return;
+    }
+    if (status === 200 && data) {
+      let sitesByDomain = this._sitesArrayToObjectByDomain(data);
+      this.etag = sitesByDomain._etag = etag;
+      sitesByDomain._timestamp = Date.now();
+      this.cache.set("sites", sitesByDomain);
+      this._dispatchChanges(sitesByDomain);
+      this.tippyTopLastUpdated = sitesByDomain._timestamp || Date.now();
+    }
+  }
+
+  _sitesArrayToObjectByDomain(sites) {
+    let sitesByDomain = {};
+    for (const site of sites) {
+      // The tippy top manifest can have a url property (string) or a
+      // urls property (array of strings)
+      for (const url of site.url ? [site.url] : site.urls || []) {
+        sitesByDomain[getDomain(url)] = site;
+      }
+    }
+    return sitesByDomain;
+  }
+
+  _dispatchChanges(data) {
+    const actionType = this.tippyTopLastUpdated ? at.TIPPYTOP_UPDATED : at.TIPPYTOP_INIT;
+    this.store.dispatch(ac.BroadcastToContent({type: actionType, data}));
+  }
+
+  async loadCachedData() {
+    const data = await this.cache.get("sites");
+    if (data && data._timestamp) {
+      this._dispatchChanges(data);
+      this.tippyTopLastUpdated = data._timestamp;
+      this.etag = data._etag;
+    }
+  }
+
+  async loadFromURL(url, headers) {
+    let data = [];
+    let etag;
+    let status;
+    try {
+      let response = await fetch(url, {headers});
+      status = response.status;
+      if (status === 200) {
+        data = await response.json();
+        etag = response.headers.get("ETag");
+      }
+    } catch (error) {
+      Cu.reportError(`Failed to load tippy top manifest from ${url}`);
+    }
+    return {data, etag, status};
+  }
+
+  async fetchIcon(url) {
+    if (!this.initialized) {
+      this.fetchQueue.push(url);
+      if (!this.initializing) {
+        await this.init();
+      }
+      return;
```

Comment:
All of this initialization/queue stuff seems a bit complex when `async fetchIcon` can just naturally queue up by `await`ing. Maybe something like:
```js
async fetchIcon(url) {
  const sitesByDomain = await this.getSitesByDomain();
  if (domain in sitesByDomain) …
}
getSitesByDomain() {
  // return an already loaded object or a promise for that object
  return this._sitesByDomain || (this._sitesByDomain = new Promise(async resolve => {
    await this.loadCachedData();
    await this.maybeRefresh();
    resolve(this._sitesByDomain); // NB one of the above set an object to this._sitesByDomain
  }));
}
```
When `this._sitesByDomain`…
- … is an `Object`, it's effectively `initialized`
- … is a `Promise`, it's effectively `initializing`
- … is `undefined`, it's basically doing `init`

And each call to `fetchIcon` can just `await this.getSites` to get the current sites and happen to wait for initialization without realizing it's initialized / initializing or not.

And however we end up `refresh`ing, it should assign the new sites object to `this._sitesByDomain`, and `fetchIcon` calls will just get the updated data.

Reply:
ooooh, the same trick we used in the `PersistentCache`. Some day I'll start thinking in promises 😉